### PR TITLE
KAS_GET_TAKES_OWNERSHIP

### LINKED ISSUE #176: Add open flag to transfer ownership of array buffers
When kastore_gets returns an array buffer it's giving back a pointer that is owned by the kastore. After #175, all array pointers returned will be individually malloced buffers, which client code may wish to take ownership of to avoid extra copies.

We need to add a flag to kas_open() which indicates that a call to ``kastore_get`` involves the transfer of ownership of the buffer from kastore to the client. Once an item has been successfully returned to the client we set a flag on the struct saying "don't free this buffer", so that fields that are not retrieved don't result in a memory leak.

What do we call this flag? Say, ``KAS_GET_OWNS_ARRAY``?

### PR DESCRIPTION
**Pull request by benjeffery**
Fixes #176

### CODE REVIEW

File: c/kastore.h [hunk 1]
```diff
@@ -94,7 +94,8 @@ Unknown flags were provided to open.
 /** @} */
 
 /* Flags for open */
-#define KAS_READ_ALL            1
+#define KAS_READ_ALL            (1 << 0)
+#define KAS_GET_TRANSFERS_OWNERSHIP            (1 << 1)
```
Comment (jeromekelleher) [inline c/kastore.h:98]: Might as well standardise on the "take" language, so how about "KAS_GET_TAKES_OWNERSHIP"

**Review by jeromekelleher — APPROVED** [review]
Awesome, looks great. Two small naming suggestions.

**Comment by benjeffery** [comment]
Great suggestions, all fixed up!

File: c/kastore.c [hunk 2]
```diff
@@ -653,14 +655,18 @@ kastore_close(kastore_t *self)
             /* We only alloc memory for the keys and arrays in write mode */
             for (j = 0; j < self->num_items; j++) {
                 kas_safe_free(self->items[j].key);
-                kas_safe_free(self->items[j].array);
+                if (!self->items[j].ownership_taken) {
```
Comment (jeromekelleher) [inline c/kastore.c:658]: Ooh, hold on - why are we checking this in write mode? This should only apply in read mode, so we can bug_assert on !ownership_taken.

File: c/kastore.h [hunk 3]
```diff
@@ -225,6 +227,11 @@ KAS_READ_ALL
     open time. This will give slightly better performance as the file can
     be read sequentially in a single pass.
 
+KAS_GET_TRANSFERS_OWNERSHIP
```
Comment (jeromekelleher) [inline c/kastore.h:230]: wrong name here

**Comment by jeromekelleher** [comment]
Spotted a couple of other things.

We should think about the "write" mode semantics too - can you open an issue to track this please? Do we currently take a copy on "put", and can we usefully avoid that when writing out a table collection in tskit?

**Comment by benjeffery** [comment]
Oh good spot, I was thinking that something could be `put` and then `get`, but now I see those are exclusive modes.

**Comment by benjeffery** [comment]
WOAH! Don't merge! Just realised nastyness happens if you read twice, kastore gives out a pointer it doesn't own.

**Comment by benjeffery** [comment]
I think the simplest fix is to say that if you use KAS_GET_TAKES_OWNERSHIP you can only read each array once?

**Comment by benjeffery** [comment]
Actually I think setting kastores array to NULL will make it realloc on a second read, will try that and add a test.

**Comment by benjeffery** [comment]
@jeromekelleher I've added a new commit here which I think is better, as when ownership is transferred we set kastore to the same state as if the column hasn't been read. This means kastore isn't giving out memory it doesn't own.

File: c/kastore.h [hunk 4]
```diff
@@ -225,6 +248,11 @@ KAS_READ_ALL
     open time. This will give slightly better performance as the file can
     be read sequentially in a single pass.
 
+KAS_GET_TAKES_OWNERSHIP
+    If this option is specified, all ``get`` operations will transfer
+    ownership of the array to the caller. ``kastore`` will not ``free``
+    the array memory and this is the responsibility of the caller.
```
Comment (jeromekelleher) [inline c/kastore.h:254]:
```suggestion
    the array memory and this is the responsibility of the caller. 
    If ``get`` is called on the same key multiple times, a new buffer will be 
    returned each time. Note that second and subsequent ``get`` calls 
    on a given key will result in ``seek`` operations even when the 
    KAS_READ_ALL flag is set, and will therefore fail on unseekable 
    streams.
```

**Review by jeromekelleher — APPROVED** [review]
Yes, this is much nicer!

I've added some documentation to clarify what happens when we have multiple gets, but good to merge then I think.

File: c/kastore.h (hunk 4 above)
Comment (benjeffery) [reply c/kastore.h:254]: Yes, good point, added.

**Comment by benjeffery** [comment]
Ok, think I'm happy now!

**Pull request merged**